### PR TITLE
Sampler, LR fix and tests for sampler and models

### PR DESCRIPTION
**Pull request by ampudia19**
I still have a handful of tests I plan to write, but I can add those in the next PR. Also note I pulled other changes elsewhere, so that I could work with the updated continuous ThompsonSampler.

### CODE REVIEW

**Comment by clinton0313** [comment]
I think what I will do is hopefully finishi running the Bayesian optimisation tests overnight/tomorrow and then put in a PR and merge those in first since those will have since been updated from when you merged them into your branch. And then this will allow me to just reject all of those files when merging this branch in to avoid dealing with the merge conflict that would arise.

File: tests/util/models/keras/models.py [hunk 1]
```diff
@@ -78,3 +81,53 @@ def trieste_deep_ensemble_model(
     model = DeepEnsemble(keras_ensemble, optimizer_wrapper, bootstrap_data)
 
     return model, keras_ensemble, optimizer_wrapper
+
+def trieste_dropout_network_model(
+    example_data: Dataset,
+    rate: Sequence[float | int] | float | int = 0.1,
```
Comment (clinton0313) [inline tests/util/models/keras/models.py]: rate has been changed to only require float now.

Comment (clinton0313) [inline tests/util/models/keras/models.py]: rate has been changed to only require float now.

File: tests/util/models/keras/models.py [hunk 2]
```diff
@@ -78,3 +81,53 @@ def trieste_deep_ensemble_model(
...
+            {"units": 300, "activation": "relu"},
+        ],
+        rate=rate
+    )
+
+    return dropout_network
+
+def trieste_mcdropout_model(
+    example_data:Dataset, 
+    rate: Sequence[float | int] | float | int = 0.1,
```
Comment (clinton0313) [inline tests/util/models/keras/models.py]: rate should only be type hinted as float.

File: trieste/models/keras/models.py [hunk 3]
```diff
@@ -335,13 +341,54 @@ def optimize(self, dataset: Dataset) -> None:
...
+
+    Note that currently we do not support setting up the model with dictionary configs and saving
+    the model during Bayesian optimization loop (``track_state`` argument in
+    :meth:`~trieste.bayesian_optimizer.BayesianOptimizer.optimize` method should be set to `False`).
+    """
     def __init__(
         self,
         model: DropoutNetwork,
         optimizer: Optional[KerasOptimizer] = None,
         num_passes: int = 200,
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: I found a default of 100 passes works well enough. Could add a comment in the docstring that increasing passes would increase the accuracy of the model at the cost of computational burden or something to that effect

File: trieste/models/keras/models.py [hunk 4]
```diff
@@ -369,8 +416,13 @@ def __init__(
         )
 
         self.num_passes = num_passes
+        self.learning_rate = learning_rate
```
Comment (clinton0313) [inline trieste/models/keras/models.py]: Won't this overwrite the learning rate that is passed in the optimizer when initializing the model?

File: trieste/models/keras/sampler.py [hunk 5]
```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
...
+        Efficiently resample network indices, and optionally quantiles, in-place without retracing.
+        """
+        self._indices.assign(self._model.sample_index(self._batch_size))  # [B]
+        if self._use_samples:
+            self._seeds.assign(tf.random.uniform(shape=(self._batch_size, 2), minval=1, maxval=999999999, dtype=tf.int32))
+
+
+class DropoutTrajectorySampler(TrajectorySampler[TrainableProbabilisticModel]):
+    """
+    This class builds functions that approximate a trajectory by using the predicted means
```
Comment (clinton0313) [inline trieste/models/keras/sampler.py:164]: maybe we should add the from :class:`~trieste.models.keras.MCDropout` thing in the docstring here?

File: trieste/models/keras/sampler.py [hunk 6]
```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
...
+            This trajectory only supports batch sizes of {self._batch_size}.
+            If you wish to change the batch size you must get a new trajectory
+            by calling the get_trajectory method of the trajectory sampler.
+            """
+        )
+
+        predictions = []
+        batch_index = tf.range(0, self._batch_size, 1)
+        _ = self._model.sample(x[:1,0,:], num_samples=1) # [DAV] Somehow first seed doesn't propagate unless I've done a dummy pred before.
+        for b, seed in zip(batch_index, tf.unstack(self._seeds)):
```
Comment (clinton0313) [inline trieste/models/keras/sampler.py:255]: Is batch_index here not just the same as putting range(self._batch_size)?

File: tests/unit/models/keras/test_models_dropout.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    qp = tf.random.uniform(tf.TensorShape(query_point_shape), minval=-4, maxval=4, dtype=tf.float32)
+    obs = tf.multiply(3, qp)
+    return Dataset(qp, obs)
+
+def _get_linear_data2(query_point_shape: ShapeLike) -> Dataset:
+    qp = tf.random.uniform(tf.TensorShape(query_point_shape), minval=-4, maxval=4, dtype=tf.float32)
+    obs = tf.multiply(-3, qp)
+    return Dataset(qp, obs)
+
+_RATE = 0.1
```
Comment (clinton0313) [inline tests/unit/models/keras/test_models_dropout.py]: You can make this into a @pytest.fixture:

@pytest.fixture(name="rate", params=[0.1])
def _rate_fixture(request: Any) -> float:
    return request.param

And then place the param rate in all the functions that require it

File: tests/unit/models/keras/test_models_dropout.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+@pytest.fixture(name="bootstrap_data", params=[False, True])
+def _bootstrap_data_fixture(request: Any) -> bool:
+    return request.param
+
+def _get_example_data(query_point_shape: ShapeLike) -> Dataset:
+    qp = tf.random.uniform(tf.TensorShape(query_point_shape), dtype=tf.float64)
+    obs = fnc_2sin_x_over_3(qp)
+    return Dataset(qp, obs)
+
+def _get_linear_data(query_point_shape: ShapeLike) -> Dataset:
```
Comment (clinton0313) [inline tests/unit/models/keras/test_models_dropout.py]: This could be one function and moved to the utils. Alternatively there already is a simple_quadratic function defined in trieste.objectives which might be simpler.

File: tests/unit/models/keras/test_models_dropout.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+    pred2, _ = model.predict(qp)
+
+    assert np.abs(pred1-pred2) > 1
+
+
+@random_seed
+@pytest.mark.parametrize("epochs", [5, 15])
+@pytest.mark.parametrize("learning_rate", [0.01, 0.1])
+@pytest.mark.parametrize("rate", [0.1, 0.2])
+def test_mcdropout_optimize(rate: int, epochs: int, learning_rate: float) -> None:
```
Comment (clinton0313) [inline tests/unit/models/keras/test_models_dropout.py]: rate should be a float here

Comment (clinton0313) [inline tests/unit/models/keras/test_models_dropout.py]: rate should be a float here

File: tests/unit/models/keras/test_sampler.py [hunk 10]
```diff
@@ -100,3 +103,77 @@ def test_ensemble_trajectory_sampler_resample_provides_new_samples_without_retra
     npt.assert_array_less(1e-4, tf.abs(evals_1 - evals_2))
     npt.assert_array_less(1e-4, tf.abs(evals_2 - evals_3))
     npt.assert_array_less(1e-4, tf.abs(evals_1 - evals_3))
+
+
+@pytest.mark.parametrize("num_evals", [10, 20])
+def test_dropout_trajectory_sampler_returns_trajectory_function_with_correctly_shaped_output(
```
Comment (clinton0313) [inline tests/unit/models/keras/test_sampler.py:109]: Should we mark these tests with pytest.mark.mcdropout as well?

**Review by clinton0313 — CHANGES_REQUESTED** [review]
Left comments with some proposed changes for you to add while still working on this branch for the next push.

File: trieste/models/keras/models.py (hunk 4 above)
Comment (ampudia19) [reply trieste/models/keras/models.py]: Mmh, this is a valid concern. How should we go about this, check if a non-default rate is set in the optimizer, otherwise set it in the model? How did your approach work around this?

File: tests/unit/models/keras/test_models_dropout.py (hunk 8 above)
Comment (ampudia19) [reply tests/unit/models/keras/test_models_dropout.py]: I'll simplify the code, but moving it to utils requires adding a bunch of extra imports there, so I'll keep it as is.

File: tests/unit/models/keras/test_models_dropout.py (hunk 9 above)
Comment (ampudia19) [reply tests/unit/models/keras/test_models_dropout.py]: good catch, thank you!

File: tests/unit/models/keras/test_sampler.py (hunk 10 above)
Comment (ampudia19) [reply tests/unit/models/keras/test_sampler.py:109]: Yes!

File: trieste/models/keras/sampler.py (hunk 5 above)
Comment (ampudia19) [reply trieste/models/keras/sampler.py:164]: Something to bring up with Hrovje, but I would agree with it.

File: trieste/models/keras/sampler.py (hunk 6 above)
Comment (ampudia19) [reply trieste/models/keras/sampler.py:255]: Yes, but I don't see how it makes a difference

**Comment by ampudia19** [comment]
I implemented most suggested changes. Thanks!

File: trieste/models/keras/models.py (hunk 4 above)
Comment (clinton0313) [reply trieste/models/keras/models.py]: You could just read the learning rate from the optimizer regardless if it is default or not and save that rate.

File: tests/unit/models/keras/test_sampler.py [hunk 11]
```diff
@@ -100,3 +103,77 @@ def test_ensemble_trajectory_sampler_resample_provides_new_samples_without_retra
...
+
+    model, _, _ = trieste_mcdropout_model(example_data, dropout=DropoutNetwork)
+
+    sampler = DropoutTrajectorySampler(model)
+    trajectory = sampler.get_trajectory()
+
+    assert trajectory(test_data).shape == (num_evals, 2)
+
+
+def test_dropout_trajectory_sampler_returns_deterministic_trajectory() -> None:
```
Comment (clinton0313) [inline tests/unit/models/keras/test_sampler.py:125]: Leaving a note here to remind about testing the conintuous thompson sampler in the bayesian optimization integration test.

**Comment by hstojic** [comment]
its a bit difficult to review this one (66 files) as it seems you have merged in latest trieste here, but it hasnt been done on the target branch

**Comment by ampudia19** [comment]
I assume that was my doing, as i worked towards amending the sampler to
take into account the new N x B x d inputs. Anything we can do on our end
to facilitate the work?

David

Le jeu. 14 avr. 2022 à 14:44, Hrvoje Stojic ***@***.***> a
écrit :

**Comment by hstojic** [comment]
did you merge in smth from original trieste repo?
or perhaps, what was the branch you were on when you created a new branch? it should have had been the target branch 
[clinton_david/mcdropout](https://github.com/clinton0313/trieste/tree/clinton_david/mcdropout)

**Comment by hstojic** [comment]
in the meantime perhaps you can tell me what files you were working on, just the sampler.py? any test file?

**Comment by ampudia19** [comment]
I merged your de-sampler update branch into mine, to account for changes
beyond our code re. the sampler. I was then mostly working on the sampler,
model tests and sampler tests these last few days.
